docs/ImpactAnalysis.md

```diff
@@ -414,8 +414,65 @@ The following experiments each pertain to several of the risks above.
+
+- **REQ-LedgerNewEra**: The ledger must be prepared with a new era that includes all changes required by CIP-164.
+
+For the remainder of this document, let's assume the changes will go into the `Euler` era, where `Conway` is currently the latest and `Dijkstra` is in preparation at the time of writing.
+
+## New block structure
+
+In Praos, all transactions to be verified and applied to the ledger state are included directly in the block body. In Leios, ranking blocks (RB) may not include transactions directly, but instead certificate and reference to an endorsement block (EB) that further references the actual transactions. This gives rise to the following requirements:
+
+- **REQ-LedgerResolvedBlockValidation**: When validating a block body, the ledger must be provided with the endorsed transactions to be applied to the ledger state.
+- **REQ-LedgerTxValidationUnchanged**: The actual transaction validation logic should remain unchanged, i.e., the ledger must validate each transaction as it does today.
+
+The endorsement block itself does not contain any additional information besides a list of transaction identifiers (hashes of the full transaction bytes). Hence, the list of transactions is sufficient to reconstruct the EB body and verify the certificate contained in the RB. The actual transactions to be applied to the ledger state must be provided by the caller of the ledger interface, typically the storage layer or the mempool.
+
+## Certificate verification
+
+In order to verify certificates contained in ranking blocks, the ledger must be aware of the voting committee and able to access their public keys. As defined in by **REQ-RegisterBLSKeys**, SPOs must be able to register their BLS keys to become part of the voting committee. The ledger will need to be able to keep track of the registered keys and use them to do certificate verification. Besides verifiying certificates, individual votes must also be verifiable by other components (e.g. to avoid diffusing invalid votes). This gives rise to:
+
+- **REQ-LedgerStateVotingCommittee**: The Leios voting committee must be part of the ledger state, updated on epoch boundaries and queryable via the state query interface.
+
+Being part of the ledger state, the voting committee will be stored in ledger snapshots and hence on disk in course of Ledger-HD. Depending on how exactly keys will be registered, the ledger might need to be able to access block headers in order to read the BLS public keys from the operational certificate. As this is not the case today (only block bodies are processed by the ledger), this results in requirement:
+
+- **REQ-LedgerBlockHeaderAccess**: The ledger must be able to access block headers.
```

Comment:
At this point, this feels too leaky to me.

How about a different approach, in which the Ledger is parameterized over a "certificate validation function" that Consensus provides to the ledger whenever asking the ledger to validate a block?

Reply:
Doable. Not going to specify functions in this document, but we could summarize this as:

- **REQ-LedgerCertificateVerification**: The ledger must be provided with a means to verify certificates in a given block body.

And amending the explanation above/below accordingly?

Reply:
We could do this like the KES verification, where we basically pass the information from the Ledger. It seems quite natural to e.g. expand the scope of the [OCERT STS](https://github.com/IntersectMBO/ouroboros-consensus/blob/main/docs/agda-spec/src/Spec/OperationalCertificate.lagda). We could pass the BLS key together with the KeyHash for the pool.

Reply:
@ch1bo I made this diagram a while ago https://ouroboros-consensus.cardano.intersectmbo.org/docs/references/block_diagrams_of_data/, I find it difficult to read, but easier than extrapolating from CDDL etc.

Last time I looked, I realized it might help to also list to various invariants that are encoded with colors here as explicit formulae.

Reply:
in that diagram, `ocertSigma` is `hbVk` signing the little triplet of `ocert*` fields.

Reply:
One of those triplet fields is `ocertN`. So it's ultimately `hbVk` that's authorizing the sequence number. (An illuminating related fact is that SPOs don't _actually_ have to increment their sequence number if they're rotating the KES key exactly when it expires.)

Reply:
That block diagram is good stuff (and would be great [cardano-blueprint](https://cardano-scaling.github.io/cardano-blueprint/) material!).

Is there a concrete suggestion for this part? Is the note I have added since to this section good enough to get this merged?